src/Forms/SearchableDropdownTrait.php

```diff
@@ -0,0 +1,594 @@
+            //
+            // Page EditForm $values for non-multi field will be
+            // [
+            //   0 => '10',
+            // ];
+            // FormBuilder $values for non-multi field will be
+            // [
+            //   'label' => 'MyTitle15', 'value' => '10'
+            // ]
+            if (array_key_exists('value', $arr)) {
+                $arr = [$arr['value']];
+            }
+            // Page EditForm $values for multi will be
+            // [
+            //   0 => '10',
+            //   1 => '15'
+            // ];
+            // FormBuilder $values for multi will be
+            // [
+            //   0 => ['label' => '10', 'value' => 'MyTitle10', 'selected' => false],
+            //   1 => ['label' => '15', 'value' => 'MyTitle15', 'selected' => false]
+            // ];
+            if (array_key_exists(0, $arr) && is_array($arr[0]) && array_key_exists('value', $arr[0])) {
+                $arr = array_map(fn ($a) => $a['value'], $arr);
+            }
+            $arr = array_filter($arr);
+            return array_map(fn ($a) => (int) $a, $arr);
+        }
+        try {
+            // things stdClass cannot be cast to string
+            (string) $value;
+        } catch (Error $e) {
+            $value = 0;
+        }
+        if (ctype_digit((string) $value) && $value != 0) {
+            return [(int) $value];
+        }
+        if ($value instanceof SS_List) {
+            return array_filter($value->column('ID'));
+        }
+        if ($value instanceof DataObject && $value->exists()) {
+            return [$value->ID];
+        }
+        // Don't know what value is, handle gracefully
+        return [];
+    }
+
+    public function Field($properties = []): DBHTMLText
+    {
+        // The entwine class needs to be added so that it is rendered in an entwine context
+        $this->addExtraClass('entwine');
+        return $this->customise($properties)->renderWith(self::class);
+    }
+
+    public function saveInto(DataObjectInterface $record): void
+    {
+        if (empty($record)) {
+            return;
+        }
+        $name = $this->getName();
+        $ids = $this->getValueArray();
+        if (substr($name, -2) === 'ID') {
+            // has_one field
+            $record->$name = $ids[0] ?? 0;
+            $record->write();
+        } else {
+            // has_many / many_many field
+            if (!method_exists($record, 'hasMethod')) {
+                throw new LogicException('record does not have method hasMethod()');
+            }
+            /** @var DataObject $record */
+            if (!$record->hasMethod($name)) {
+                throw new LogicException("Relation $name does not exist");
+            }
+            /** @var Relation $relation */
+            $relation = $record->$name();
+            if (!is_a($relation, Relation::class)) {
+                throw new LogicException("$name is not a Relation");
+            }
```

Comment:
```suggestion

            /** @var DataObject $record */
            if (!$record->hasMethod($name)) {
                throw new LogicException("Relation $name does not exist");
            }
            
            /** @var Relation $relation */
            $relation = $record->$name();
            if (!is_a($relation, Relation::class)) {
                throw new LogicException("$name is not a Relation");
            }
            
```
Not tooo fussed about this since we don't have specific coding standards about line breaks around blocks.